README.md
# onEVENT

onEVENT is an event based automation tool for Linux. It watches events such as battery percentage, input devices, cpu and even facebook notifications. You can specify any Terminal command to be run when these event results are what you want them to be.

## Command line arguments

This program supports a few command line params. None of these are needed to run the program.

- --verbose -v		: Verbose mode. Tells the program to output information to the terminal. It uses my pprint lib which makes things look purrdy :3
- --folder		: The folder the event files are located in. Defaults to events/
- --file		: The JSON file that your events are in. Defaults to events.json
- --timeout		: The delay between checking all the events. Defaults to 1.
- --server -s		: Makes onEVENT send data through sockets.
- --host -o		: The hostname to bind on. Defaults to ''
- --port -p		: The port to bind on. Defaults to 9987
- --local -l		: Makes it run on local mode. Not default.

## Event file

onEVENT uses a 3rd party library for YAML parsing. If you don't have this, it uses my old parser.

When you have PyYAML

```YAML
- onEVENT:
    repeat: 0
    delay:
        seconds: 1
    when:
        - procexists:
            params: 
                - banshee
            result: True
    do:
        - action:
            - notify-send
            - Banshee open
            - Lowering your volume so you don't get blasted with sound!
        - action:
            - amixer
            - set
            - Master
            - 60%
    alternative:
        - action:
            - notify-send
            - Banshee closed
            - Raising volume back to 100%
        - action:
            - amixer
            - set
            - Master
            - 100%
```

**Here is how it works**

- repeat - Specify 1 if you want this to repeat over and over, even if the state doesn't go from True to false or False to True.
- delay - The delay between running of events. Can be a combonation of seconds minutes or hours.
- when - Events to run
- params - The params for the event.
- result - What you want the event to be. True / False / 1 / 0
- do - The actions
- action - The linux command to run. Each param is seperated by newline or like [param1, param2, param2]. You can specify more than one of these.

( optional )

- alternative - The alternative actions to run.
- action - The linux command to run when the event switches to false. You can specify more than one of these.


When you don't have PyYAML

```YAML
battery(full) = 1:
    repeat: 0
    delay: {'seconds': 0}
    action: [['notify-send', 'Battery', 'Battery is fully charged.']]
``` 

**Here is how it works**

event(params) = result:
    repeat: 0
    delay: {'seconds': 0}
    action: [['some', 'action']]

- event is the event name.
- params is a comma seperated list of params
- result is what state you want the event to be in to do the action.

- repeat - Specify 1 if you want this to repeat over and over, even if the state doesn't go from True to false or False to True.
- delay - The delay in between the event's run.
- action - The action to do when all events meet their result. 

( optional )

- alternative - The action to do when one of the events don't meet their result.


Each action should be seperated like the following:

[['actioncommand', 'param1', 'param2'], ['another_action', '-t', 'param for -t']]    

### Combining events

You can combine events. It's really simple.

```YAML
event1(params) = 1 & event2(params) = 0:
```

## Built in events

This is a list of events i have built, and plan to build. Feel free to contact me with any ideas or suggestions for events, i will be happy to hear them and maybe even implement them!

- [X]	Battery
- [X]	Brightness
- [X]	CPU Percentage
- [ ]	Email
- [X]	Facebook Notification ( Outdated. Use the RSS event now :) )
- [X]	File changed
- [X]	Filepath Exists
- [X]	Input Device plugged in
- [X]	Internet connection
- [X]	Lid closed ( Laptop )
- [X]	New file in directory
- [X]	Process exists
- [X]	Power button press
- [X]	RAM Usage
- [X]	RSS
- [ ]	Temperature
- [X]	Time
- [X]	Uptime
- [ ]	Many more!


**notes**

Some of these commands load from files that may be different on your system. If you experience an error like 'FileNotFound' or something similar you may need to change some code. If you're not familiar with Python this may not be easy. 

Please pay attention to what file the error is coming from then search for the linux file it may be. For example, if the battery event puts out an error google 'linux ( your distro ) battery class file' or something similar. 

You will then need to change the directory in the event file. This should be easy to locate, since only 1 part of my code should look like a file directory.


####Battery
The battery percentage may not be the same as your monitor one is. This does not mean its wrong, i have no idea how the system monitors it.

I might be doing the math wrong, but i've checked multiple sources and they all they thats how you do it.

####Facebook Notification
Yes, you heard right! This program supports Facebook Notifications. This runs through the RSS event now. So make sure to use that one!

1. Open your facebook to https://www.facebook.com/notifications
2. Click the little RSS button. This is next to the 'Get Notifications via Text message'.
4. Paste the link into your event param, the next param can be anything, just make it different since it uses this as the name of the file that stores the previous RSS info. So something like 'Facebook_Feed'.

####RAM

The ram does not find the 'usable' amount. It just goes off the max amount your PC has. This means it might be different that either your system monitor, or other programs such as conky.


## Adding your own events

Say what? I can add my own events?!?
Yes, you certainly can. It is very easy.

Just make a new .py file in the events folder, name it whatever you want.
Inside the file define a function that is the same as your file name. For example if i want, say a weather event. Ill name it weather.py and put in

```Python
def weather(params):
	#code
	return (1, 'Output')
```

Now, you may be thinking how my program knows if this event will be true or not. You have to return a tuple in each function.

```Python
return (1, 'Param1', 'param2')
```
or

```Python
return (0, 'Param1', 'Param2')
```

If the first item is 1 the event returns true, 0 is false. You can also return params, which can be used in the 'action' in the events file.

For example, if i return (1, 'This is a test!')
and in the action i use {0}, {0} will get replaced with This is a test!
{0} is the first param you return, {1} is the second, and so on.
You can specify more params and use {1}, {2}, {3}  to your desire!
Fun huh? :D

## Examples

Here are a few quick examples of events you could set up.

When you have the YAML library.

```YAML
- onEVENT:
    repeat: 0
    delay:
        seconds: 1
    when:
        - newfile:
            params:
                - /media/steven/
            result: True
    do:
        - action:
            - notify-send
            - New mount
            - A new directory has been mounted "{0}"
```

When you don't have the YAML library.

Notifies me when my battery is charging.

```YAML
battery(charging) = 1:
    repeat: 0
    delay: {'seconds': 0}
    action: [['notify-send', 'Battery', 'Battery is charging ({0}%)']]
    alternative: [['notify-send', 'Battery', 'Battery is discharging. ({0}%)']]
```

Notifies me when my battery is full.

```YAML
battery(full) = 1:
    repeat: 0
    delay: {'seconds': 0}
    action: [['notify-send', 'Battery', 'Battery is fully charged.']]
```

Notifies me when i plug / unplug my hard drive.

```YAML
exists(/media/steven/External) = 1:
    repeat: 0
    delay: {'seconds': 0}
    action: [['notify-send', 'External HD', 'External HD has been plugged in.']]
    alternative: [['notify-send', 'External HD', 'External HD has been un-plugged.']]
```

## Server

onEVENT has the ability to send data over sockets. It sends event info in json format.

## Notes

You should be aware that some of these events might not return the same results as your system. For example, on my system, my program measures the battery level at roughly 2% different than my system does. I have no idea what my system is doing, or if my math is wrong ( multiple sources say it's right ). This should not discourage you from using this program, though, you should be aware of this little glitch. 

If you are able to make a better equation that has better results i would be very glad to see it and possibly implement it into these events.


### Root commands

If the command you have defined in "action" requires root to run, by default it will fail. You can run sudo python3 onEVENT.py to give onEVENT the ability to run root commands.


## Contact

If you wish to contact me you can use any of these methods:

- Email
	- Main 		> 42Echo6Alpha@gmail.com
	- Alternative 	> Sjc1000@hotmail.com or insertfunnyaddress@hotmail.com

- IRC
	I am on IRC quite often.
	server:	irc.freenode.net ( port 6667 by default )
	channel: #Sjc_Bot is my channel, If im Sjc_AFK i am not connected. I have a bouncer that changes to AFK when i disconnect.


## Thanks

Thank you for using my software. This is and always will be free to use. I will not charge a dime for any of my solo development projects.


changelog.md
**1.3.1**

    - Now checks for PyYAML, and if it doesn't have it. Used my custom breed of YAML.

**1.3.0**

    - Changed from JSON syntax to a custom breed of YAML

**1.2.0**
    
    - Added folderchanged event. Watches all the files modify time and returns true when one has been changed.
    - Added universal error handler. If any error happens during the running of
        event it will report the error. It will also skip the last data section, but not the last time section.
        This means it will wait for the specified time until running it again.
    - Added server capability, you can now run onEVENT as a server.
    - Bugfixes

**1.1.2**

    - Added in an error handling to onEVENT,
        if an event has an error itl return -1
        when it does onEVENT will skip the last data section.
    - Added an optional iterate param for the event.json file.
    - Removed Facebook event. Now use the RSS event.
    - Changed Time event. Has more options.
    - Added more events.
    - Added GNU General Public License to onEVENT.
    - Tasks now run once straight away, then wait depending on their delay option.

**1.1.1**

    - Added more events.
    - Fixed the timing of the events.
    - Added command line options such as -v for verbose.
    - Now uses my pprint function for printing to the command line.

**1.1.0**

    - Complete rewrite of the base system. Supports multiple events better.

**1.0.0**

    - First write of the program.


onEVENT.py
#!/urs/bin/env python3
import json
import ast
import os
import imp
import time
import datetime
import threading
from subprocess import call, DEVNULL
import argparse
import sys
import socketserver
from cprint import cprint

try:
    old_parser = False
    import eventparser as eventparser
except ImportError:
    old_parser = True
    import old_eventparser as eventparser

"""
    onEVENT is an event based automation system for Linux made by Sjc1000 
                ( Steven J. Core )
            Copyright © 2015, Steven J. Core.
    
    onEVENT is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""


__version__ = '1.3.1'


def timestamp():
    time_object = time.localtime(time.time())
    return '[{:0>2}:{:0>2}:{:0>2}]'.format(time_object[3],
              time_object[4],time_object[5])


class NewEvent():
    """newevent
    A class for handling all the event data. With a few methods. 
    Such as run and reload.
    """
    events = None
    action = None
    delay = None
    repeat = None
    last_data = None
    last_time = None
    alternative = None
    def __init__(self, events, action, repeat, delay, alternative=None,
                 id=None, iterate=0):
         
        self.events = events
        self.repeat = repeat
        
        if old_parser is False:
            self.action = action
            self.delay = delay
            if alternative is not None:
                self.alternative = alternative
        else:
            self.action = ast.literal_eval(action)
            self.delay = ast.literal_eval(delay)
            if alternative is not None:
                self.alternative = ast.literal_eval(alternative)
        self.id = id
        self.iterate = iterate
        
    def run(self, sources):
        """run
        Runs the event file.
        Loading from the sources dictionary.
        """
        output = []
        for event in self.events:
            function = getattr(sources[event['event']], event['event'])
            output.append(function(*event['params']))    
        return output


class onEVENT():
    """onEVENT class
    The base system of onEVENT.
    """
    events = []
    sources = {}
    def __init__(self, eventfolder, eventfile):
        """__init__
        params:
            - eventfolder - The folder where all the events are stored.
            - eventfile - The event file to load from.
        """
        self._eventfolder = eventfolder
        self._eventfile = eventfile
        self.loadevents()
        
        if attemptreload:
            reload_thread = threading.Thread(target=self.reloadloop)
            reload_thread.daemon = True
            reload_thread.start()
        
        if verbose:
            cprint('[.green] Now watching events!', '[.purple]' + timestamp())
        
        if server:
            self.server = socketserver.server(host, port, local)
            connect_thread = threading.Thread(
                target=self.server.listen)
            connect_thread.daemon = True
            connect_thread.start()
    
    def reloadloop(self):
        """reloadloop
        The reload loop. Started in a different thread.
        This attemps to reload event data while the program is in use.
        It will only do this if you run onEVENT -r
        """
        while True:
            time.sleep(60)
            if verbose:
                cprint('[.green] Reloading events.', '[.purple]' + timestamp())
            self.loadevents(True)
        return None
    
    def loadevents(self, forcenoverbose=False):
        """loadevents
        Loads all the events and creates event classes.
        """
        self.events = []
        with open(self._eventfile) as efile:
            edata = eventparser.parse(efile.read())

        if verbose and forcenoverbose is False:
            cprint(' Reading event file and creating event classes.',
                   '[.purple]' + timestamp())
        
        for index, ev in enumerate(edata):
            if not 'alternative' in ev:
                ev['alternative'] = None
            if not 'iterate' in ev:
                ev['iterate'] = 0
            new_event = NewEvent(
                ev['on'], ev['action'], ev['repeat'], 
                ev['delay'], ev['alternative'], index, ev['iterate'])
            
            if verbose and forcenoverbose is False:
                cprint('\t[.blue]Creating event class > [.green]' + 
                       ', '.join([x['event'] for x in ev['on']]),
                       '[.purple]' + timestamp())
            self.events.append(new_event)
        
        filelist = [f for f in os.listdir(self._eventfolder) if 
                    f.endswith('.py')]
        if verbose and forcenoverbose is False:
            cprint(' Importing event source files.',
                   '[.purple]' + timestamp())
        for event in filelist:
            if verbose and forcenoverbose is False:
                cprint('\t[.blue]Importing [.green]' + event, '[.purple]' + 
                       timestamp(), '\n[.red]')
            try:
                self.sources[event[:event.index('.')]] = imp.load_source(
                    event, self._eventfolder + event )
            except Exception as Error:
                if forcenoverbose is False:
                    cprint('[.red] Error raised while importing ' + event + ': ' + str(Error), '[.purple]' + timestamp())
        return None
    
    def checktime(self, event):
        """checktime
        Checks the events last time against the current time.
        This returns 1 if the event can be run.
        params:
            - event - The event to check.
        """
        current_time = time.time()
        c = current_time - event.last_time
        m, s = divmod(c, 60)
        h, m = divmod(m, 60)
        check_time = {'seconds': s, 'minutes': m, 'hours': h}
        for d in event.delay:
            if check_time[d] > int(event.delay[d]):
                return False
        return True
    
    def checkoutput(self, event, event_data):
        """checkoutput
        Checks the output of an event against the last event data.
        """
        for index, response in enumerate(event_data):
            if int(response) != int(event.events[index]['result']):
                return False
        return True
    
    def eventloop(self, timeout=1):
        """eventloop
        The main event loop. Runs all the events.
        """
        while True:
            for index, event in enumerate(self.events):
                event_thread = threading.Thread(
                    target=self.callevent,
                    args=(index, event))
                event_thread.daemon = True
                event_thread.start()
            time.sleep(timeout)
        return None
    
    def callevent(self, index, event):
        """callevent
        Calls the event.
        params:
            - index - The index of the event.
            - event - The event itself.
        """
        current_time = time.time()
         
        if event.last_time is not None:
            if self.checktime(event):
                return None

        self.events[index].last_time = current_time
        try:
            event_data = event.run(self.sources)
        except Exception as Error:
            if verbose:
                cprint(' [.red]Error raised in event - ' + 
                   ', '.join([x['event'] for x in event.events]) + 
                   ' - ' + str(Error), '[.purple]' + timestamp())
            return None

        check_data = [x[0] for x in event_data]
        
        if int(event_data[0][0]) == -1 and verbose:
            cprint(' [.red]Error > ' + event.events[0]['event'], 
                   '[.purple]' + timestamp())
            return None
        
        if event.last_data is None:
            self.events[index].last_data = check_data
        
        if int(event.repeat) == 0 and event.last_data == check_data:
            return None
        if verbose:
            space1 = 15-len(event.events[0]['event'])
            space = 25-len(', '.join(event.events[0]['params'])[:24])
            cprint(' [.blue]' + event.events[0]['event'][:13] + ' '*space1 + 
                   '| [.green]' + 
                   ', '.join(event.events[0]['params'])[:24] + 
                   ' ' * space + '[.blue]|[.purple][.bold] ' + 
                   str(check_data), '[.purple]' + timestamp())
        if server:
            e = event.events
            for a, b in zip(enumerate(e), event_data):
                i, a = a
                a['output'] = b
                e[i] = a
            self.server.send(json.dumps(e))
                
        params = []
        output = self.checkoutput(event, check_data)
        
        self.events[index].last_data = check_data
        
        for p in event_data:
            for i, par in enumerate(p):
                if i == 0:
                    continue
                params.append(par)
        
        if output:
            if isinstance(params[0], list) and event.iterate:
                for data in params[0]:
                    for command in enumerate(event.action):
                        if isinstance(data, dict):
                            command = [c.format(**data) for c in command[1]]
                        else:
                            command = [c.format(*data) for c in command[1]]
                        call(command, stdout=DEVNULL)
            else:
                for command in enumerate(event.action):
                    command = [c.format(*params) for c in command[1]]
                    call(command, stdout=DEVNULL)
    
        if output is False and event.alternative is not None:
            for command in enumerate(event.alternative):
                command = [c.format(*params) for c in command[1]]
                call(command, stdout=DEVNULL)
        return None


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='onEVENT event based automation system for Linux.',
        prog='onEVENT')
    parser.add_argument(
        '--file', 
        help='The location of the file where your events are.',
        default='eventfile')
    parser.add_argument(
        '--folder', 
        help='The folder where all the events are stored.',
        default='events/')
    parser.add_argument(
        '-v', '--verbose',
        help='Verbose mode. Displays the output of everything',
        action='store_true')
    parser.add_argument(
        '-r', '--reload',
        help='Makes the program attempt to reload info from the event '
        'file every 60 seconds. Will not add new events just reloads '
        'current event info.',
        action='store_true')
    parser.add_argument(
        '-V', action='version', 
        version='%(prog)s is currently version ' + str(__version__))
    parser.add_argument(
        '-T', '--timeout', help='The timeout between the checking the '
        'events.',
        default=1, type=int)
    parser.add_argument(
        '-s', '--server', help='Turns this into a server.',
        action='store_true', 
        default=False)
    parser.add_argument(
        '-o', '--host', help='The host for this to run on.', 
        type=str, default='')
    parser.add_argument(
        '-p', '--port', help='The port for this to run on.', type=int, 
        default=9987)
    parser.add_argument(
        '-l', '--local', action='store_true', help='Run it locally or '
        'online.', 
        default=False)
    parser.add_argument(
        '-e', '--showall', help='Shows every action that onEVENT makes.',
        action='store_true', default=False)

    args = parser.parse_args()
    verbose = args.verbose
    eventfile = args.file
    attemptreload = args.reload


    if args.server:
        server = True
        host = args.host
        port = args.port
        local = args.local
    else:
        server = False

    try:
        onEVENT = onEVENT(args.folder,args.file)
        onEVENT.eventloop(args.timeout)
    except KeyboardInterrupt:
        print('')
        cprint('[.red]Shutting down.')
        if server:
            onEVENT.server.shutdown()
